packages/test-base/src/commonTest/kotlin/io/realm/kotlin/test/common/RealmAnyNestedCollectionTests.kt

```diff
@@ -0,0 +1,810 @@
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertTrue
+
+class RealmAnyNestedCollectionTests {
+
+    private lateinit var configBuilder: RealmConfiguration.Builder
+    private lateinit var configuration: RealmConfiguration
+    private lateinit var tmpDir: String
+    private lateinit var realm: Realm
+
+    @BeforeTest
+    fun setup() {
+        tmpDir = PlatformUtils.createTempDir()
+        configBuilder = RealmConfiguration.Builder(
+            setOf(
+                JsonStyleRealmObject::class,
+                Sample::class,
+            )
+        ).directory(tmpDir)
+        configuration = configBuilder.build()
+        realm = Realm.open(configuration)
+    }
+
+    @AfterTest
+    fun tearDown() {
+        if (this::realm.isInitialized && !realm.isClosed()) {
+            realm.close()
+        }
+        PlatformUtils.deleteTempDir(tmpDir)
+    }
+
+    // Set
+    // - Import primitive values
+    // - Set primitive values
+    // - Notifications
+    @Test
+    fun setInRealmAny_copyToRealm() = runBlocking {
+        val sample = Sample().apply { stringField = "SAMPLE" }
+        realm.write {
+            val instance = JsonStyleRealmObject().apply {
+                // Assigning set
+                // - How to prevent adding a set containing non-any elements?
+                //   - Can we do this on the fly!?
+                value = RealmAny.create(
+                    realmSetOf(
+                        RealmAny.create(5),
+                        RealmAny.create("Realm"),
+                        RealmAny.create(sample)
+                    )
+                )
+            }
+            copyToRealm(instance)
+        }
+        val anyValue: RealmAny = realm.query<JsonStyleRealmObject>().find().single().value!!
+        assertEquals(RealmAny.Type.SET, anyValue.type)
+    }
+
+    @Test
+    fun setInRealmAny_assignment() = runBlocking {
+        realm.write {
+            copyToRealm(Sample().apply { stringField = "SAMPLE" })
+            val instance = copyToRealm(JsonStyleRealmObject())
+            // Assigning set
+            // - How to prevent adding a set containing non-any elements?
+            //   - Can we do this on the fly!?
+            instance.value = RealmAny.create(
+                realmSetOf(
+                    RealmAny.create(5),
+                    RealmAny.create(4),
+                    RealmAny.create(6)
+                )
+            )
+            instance
+        }
+        val anyValue: RealmAny = realm.query<JsonStyleRealmObject>().find().single().value!!
+        assertEquals(RealmAny.Type.SET, anyValue.type)
+    }
+
+    @Test
+    fun setInRealmAny_throwsOnNestedCollections_copyToRealm() = runBlocking<Unit> {
+        realm.write {
+            JsonStyleRealmObject(ObjectId().toString()).apply {
+                value =
+                    RealmAny.create(realmSetOf(RealmAny.create(realmListOf(RealmAny.create(5)))))
```

Comment:
Hmm, I guess this ties a bit to the input validation that we did in #1443. Maybe we should try to catch this. But this would either require checking it for any input operation in the generic implementation of `UnmanagedRealmSet`s or have some way of differentiating the implementation 🤔

Reply:
I created https://github.com/realm/realm-kotlin/issues/1511 to track the general task of enforcing realm constraints on unmanaged data.